### PR TITLE
Add link purging via Message content (regex) pattern checks

### PR DESCRIPTION
**Pull request by patrickklaeren**

### CODE REVIEW

**Review by Scott-Caldwell — DISMISSED** [review]
Blocking, since I'm not convinced that removing the invite purging behavior is a good move.

File: Modix.Bot/Modules/MessageContentPatternModule.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+using Modix.Data.Models.Core;
+using Modix.Services.Blocklist;
+
+namespace Modix.Bot.Modules
+{
+    [Name("Patterns")]
+    [Summary("Use and maintain message content patterns.")]
+    [Group("pattern")]
+    [Alias("patterns")]
+    public class MessageContentPatternModule : ModuleBase
```
Comment (Scott-Caldwell) [inline Modix.Bot/Modules/MessageContentPatternModule.cs:13]: Would it makes sense to have a "list" command? I can see it being very easy to forget what we have configured.

File: Modix.Data/Migrations/20200905130436_AddMessageContentPatterns.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                },
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_MessageContentPatterns", x => x.Id);
+                });
+
+            migrationBuilder.Sql("UPDATE public.\"ClaimMappings\" SET \"Claim\"='BypassMessageContentPatternCheck' WHERE \"Claim\"='PostInviteLink';");
+        }
+
+        protected override void Down(MigrationBuilder migrationBuilder)
```
Comment (Scott-Caldwell) [inline Modix.Data/Migrations/20200905130436_AddMessageContentPatterns.cs:28]: Should `Down` have SQL to update the claim mappings back to `PostInviteLink`?

File: Modix.Bot/Modules/MessageContentPatternModule.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        public MessageContentPatternModule(IMessageContentPatternService messageContentPatternService)
+        {
+            _messageContentPatternService = messageContentPatternService;
+        }
+
+        [Command("block")]
+        [Summary("Adds a new pattern to block.")]
+        public async Task BlockAsync([Summary("Regex pattern for the blocked content."), Remainder] string pattern)
+        {
+            if (await _messageContentPatternService.DoesPatternExist(Context.Guild.Id, pattern))
```
Comment (Scott-Caldwell) [inline Modix.Bot/Modules/MessageContentPatternModule.cs:26]: Somewhat minor, but `AddPattern` is already doing this check internally, so we're querying the DB more often than we have to for this command. Would it make sense to just let the service do this check?

File: Modix.Data/Models/Core/MessageContentPatternEntity.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+        [Required]
+        public string Pattern { get; set; }
+
+        public MessageContentPatternType PatternType { get; set; }
+
+        public ulong GuildId { get; set; }
+    }
+
+    public class MessageContentPatternEntityConfiguration
```
Comment (Scott-Caldwell) [inline Modix.Data/Models/Core/MessageContentPatternEntity.cs:23]: Might make sense to have a unique constraint on (GuildId, Pattern) to ensure we can't have bad data

File: Modix.Services/Moderation/MessageContentCheckBehaviour.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            if (await _authorizationService.HasClaimsAsync(author, AuthorizationClaim.BypassMessageContentPatternCheck))
+            {
+                Log.Debug("Message {MessageId} was skipped because the author {Author} has the {Claim} claim",
+                    message.Id, message.Author.Id, AuthorizationClaim.BypassMessageContentPatternCheck);
+                return;
+            }
+
+            Log.Debug("Message {MessageId} is going to be deleted", message.Id);
+
+            await _moderationService.DeleteMessageAsync(message, "Unauthorized Message Content Link",
```
Comment (Scott-Caldwell) [inline Modix.Services/Moderation/MessageContentCheckBehaviour.cs:86]: Minor, but these aren't necessarily links anymore.

```suggestion
            await _moderationService.DeleteMessageAsync(message, "Unauthorized Message Content",
```

File: Modix.Services/Moderation/MessageContentCheckBehaviour.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                _discordSocketClient.CurrentUser.Id, default);
+
+            Log.Debug("Message {MessageId} was deleted because it contains blocked content", message.Id);
+
+            await message.Channel.SendMessageAsync($"Sorry {author.Mention} your link has been removed!");
```
Comment (Scott-Caldwell) [inline Modix.Services/Moderation/MessageContentCheckBehaviour.cs:91]: Same as above, should probably reword not to mention links.

File: Modix.Services/Moderation/MessageContentCheckBehaviour.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        }
+
+        private async Task<bool> IsContentBlocked(IGuildChannel channel, IMessage message)
+        {
+            var patterns = await _messageContentPatternService.GetPatterns(channel.GuildId);
```
Comment (Scott-Caldwell) [inline Modix.Services/Moderation/MessageContentCheckBehaviour.cs:96]: I'm hoping this won't be too bad, doing an extra database call every time someone posts a message. A good future optimization could be to cache the patterns in memory so we don't have to pull them constantly.

File: Modix.Services/Moderation/MessageContentCheckBehaviour.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+                    var regex = new Regex(
+                        pattern: pattern,
+                        options: RegexOptions.IgnoreCase,
+                        matchTimeout: TimeSpan.FromSeconds(2));
+
+                    var blockMatches = regex.Matches(message.Content);
+
+                    return blockMatches.Any();
+                }
+            }
```
Comment (Scott-Caldwell) [inline Modix.Services/Moderation/MessageContentCheckBehaviour.cs:129]: This logic will allow a message that contains blocked content, so long as it also contains allowed content in at least one other part of the message.

For example, if we block `discord.gg/robloxexploits` and allow `discord.gg/csharp`, this message will be allowed:
```
Visit discord.gg/robloxexploits and discord.gg/csharp!
```

Instead, I think we want to loop over the specific parts of the message that match the blocked regex and check whether they're allowed.

**Review by Scott-Caldwell — CHANGES_REQUESTED** [review]
Requesting changes this time because there's some faulty logic in the behavior that will unintentionally allow blocked content

File: Modix.Services/Moderation/MessageContentCheckBehaviour.cs (hunk 8 above)
Comment (Scott-Caldwell) [reply Modix.Services/Moderation/MessageContentCheckBehaviour.cs:129]: We should add a unit test for this scenario as well

File: Modix.Data/Migrations/20200905130436_AddMessageContentPatterns.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                    Pattern = table.Column<string>(nullable: false),
+                    PatternType = table.Column<string>(nullable: false),
+                    GuildId = table.Column<long>(nullable: false)
+                },
+                constraints: table =>
+                {
+                    table.PrimaryKey("PK_MessageContentPatterns", x => x.Id);
+                });
+
+            migrationBuilder.Sql("UPDATE public.\"ClaimMappings\" SET \"Claim\"='BypassMessageContentPatternCheck' WHERE \"Claim\"='PostInviteLink';");
```
Comment (Scott-Caldwell) [inline Modix.Data/Migrations/20200905130436_AddMessageContentPatterns.cs:25]: If we're going to delete the invite purging behavior, I think we'll want to include the invite link regex as part of the migration here. Other guilds also use Modix, so I'd prefer not to break the invite link deletion functionality for them.

File: Modix.Bot/Modules/MessageContentPatternModule.cs (hunk 3 above)
Comment (patrickklaeren) [reply Modix.Bot/Modules/MessageContentPatternModule.cs:26]: Yes, absolutely. Had this in here before I added `ServiceResponse`

File: Modix.Services/Moderation/MessageContentCheckBehaviour.cs (hunk 8 above)
Comment (patrickklaeren) [reply Modix.Services/Moderation/MessageContentCheckBehaviour.cs:129]: Good catch - did not test this.

**Review by Scott-Caldwell — APPROVED** [review]
The CI is failing because it needs a migration for the index.

Side note: if you block `test` and allow `testing`, it'll still block messages containing "testing". I think this should be fine for now, since it's erring on the side of safety.

LGTM after adding the migration.

**Pull request merged**